dbms/src/DataStreams/HashJoinProbeBlockInputStream.h

```diff
@@ -38,38 +37,108 @@ class HashJoinProbeBlockInputStream : public IProfilingBlockInputStream
+     *                  |                      ---------------
+     *                  |  has non_joined data |             | no non_joined data
+     *                  |                      ▼             |
+     *                  |             READ_NON_JOINED_DATA   |
+     *                  |                      \             /
+     *                  |                       \           /
+     *                  |                        \         /
+     *                  |                         \       /
+     *                  |                          \     /
+     *                  |                           \   /
+     *                  |                            \ /
+     *                  |                             ▼
+     *                  |                      GET_RESTORE_JOIN
+     *                  |                             |
+     *                  |                             ▼
+     *                  |                      ---------------
+     *                  |    has restored join |             | no restored join
+     *                  |                      ▼             ▼
+     *                  |                RESTORE_BUILD    FINISHED
+     *                  |                      |
+     *                  -----------------------|
+     *
+     */
     enum class ProbeStatus
     {
-        PROBE,
-        WAIT_FOR_READ_NON_JOINED_DATA,
-        READ_NON_JOINED_DATA,
-        FINISHED,
+        WAIT_BUILD_FINISH, /// wait build finish
+        PROBE, /// probe for both init probe and restore probe
+        WAIT_PROBE_FINISH, /// wait probe finish
+        GET_RESTORE_JOIN, /// try to get restore join
+        RESTORE_BUILD, /// build for restore join
+        READ_NON_JOINED_DATA, /// output non joined data
+        FINISHED, /// the final state
     };
-    void readSuffixImpl() override;
-    void finishOneProbe();
 
+    Block getOutputBlock();
+    std::tuple<size_t, Block> getOneProbeBlock();
+    void onCurrentProbeDone();
+    void onAllProbeDone();
+    void onCurrentReadNonJoinedDataDone();
+    void tryGetRestoreJoin();
+    void readSuffixImpl() override;
     const LoggerPtr log;
+    /// join/non_joined_stream/restore_build_stream/restore_probe_stream can be modified during the runtime
+    /// although read/write to those are almost only in 1 thread, but an exception is cancel thread will
+    /// read them, so need to protect the multi-threads access
+    std::mutex mutex;
+    JoinPtr original_join;
     JoinPtr join;
-    size_t probe_index;
+    const bool need_output_non_joined_data;
+    size_t current_non_joined_stream_index;
+    BlockInputStreamPtr current_probe_stream;
+    UInt64 max_block_size;
     ProbeProcessInfo probe_process_info;
     BlockInputStreamPtr non_joined_stream;
-    ProbeStatus status{ProbeStatus::PROBE};
+    BlockInputStreamPtr restore_build_stream;
+    BlockInputStreamPtr restore_probe_stream;
```

Comment:
how about replacing by `RestoreInfo`?

Reply:
I think it's better not to expose `restore_join`, `restore_stream` to `HashJoinProfbeBlockInputStream`.
The `HashJoinProfbeBlockInputStream` will only see the `RestoreInfo` and will only operate on the `RestoreInfo`

Reply:
Like `RestoreInfo::restoreBuild` and `RestoreInfo::restoreProbe`?